File: Study Plans/Algorithms 1/278FirstBadVersion.cpp
// The API isBadVersion is defined for you.
// bool isBadVersion(int version);
#include <stdlib.h>
#include <stdio.h>
using namespace std;

class Solution {
public:
    int firstBadVersion(int n) 
    {
        int left = 1;
        int right = n;
        int mid;

        // While the left and right pointers don't cross
        while (left < right)
        {
            int midIndex = left + (right - left) / 2;
            // If the middle value is bad, return it
            if (isBadVersion(midIndex))
            {
                right = midIndex;
            }
            else
            {
                // If the middle value is good, move the left pointer to the right to continue checking
                left = midIndex + 1;
            }
        }
        return left;
    }
};

File: Study Plans/Binary Search/278FirstBadVersion.java
public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int min = 1, max = n, mid;

        while (min <= max)
        {
            mid = min + ((max - min) / 2);

            // If version is not bad, move up
            if (!isBadVersion(mid))
                min = mid + 1;
            else
                max = mid - 1;
        }
        return min;
    }
}